// @adiwajshing/baileys-md/lib/Utils/auth-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSingleFileAuthState = exports.initAuthCreds = exports.initInMemoryKeyStore = void 0;
const crypto_1 = require("crypto");
const WAProto_1 = require("../../WAProto");
const crypto_2 = require("./crypto");
const generics_1 = require("./generics");
const initInMemoryKeyStore = ({ preKeys, sessions, senderKeys, appStateSyncKeys, appStateVersions } = {}, save) => {
    preKeys = preKeys || {};
    sessions = sessions || {};
    senderKeys = senderKeys || {};
    appStateSyncKeys = appStateSyncKeys || {};
    appStateVersions = appStateVersions || {};
    const keyData = {
        preKeys,
        sessions,
        senderKeys,
        appStateSyncKeys,
        appStateVersions,
    };
    return {
        ...keyData,
        getPreKey: keyId => preKeys[keyId],
        setPreKey: (keyId, pair) => {
            if (pair)
                preKeys[keyId] = pair;
            else
                delete preKeys[keyId];
            save(keyData);
        },
        getSession: id => sessions[id],
        setSession: (id, item) => {
            if (item)
                sessions[id] = item;
            else
                delete sessions[id];
            save(keyData);
        },
        getSenderKey: id => {
            return senderKeys[id];
        },
        setSenderKey: (id, item) => {
            if (item)
                senderKeys[id] = item;
            else
                delete senderKeys[id];
            save(keyData);
        },
        getAppStateSyncKey: id => {
            const obj = appStateSyncKeys[id];
            if (obj) {
                return WAProto_1.proto.AppStateSyncKeyData.fromObject(obj);
            }
        },
        setAppStateSyncKey: (id, item) => {
            if (item)
                appStateSyncKeys[id] = item;
            else
                delete appStateSyncKeys[id];
            save(keyData);
        },
        getAppStateSyncVersion: id => {
            const obj = appStateVersions[id];
            if (obj) {
                return obj;
            }
        },
        setAppStateSyncVersion: (id, item) => {
            if (item)
                appStateVersions[id] = item;
            else
                delete appStateVersions[id];
            save(keyData);
        }
    };
};
exports.initInMemoryKeyStore = initInMemoryKeyStore;
const initAuthCreds = () => {
    const identityKey = crypto_2.Curve.generateKeyPair();
    return {
        noiseKey: crypto_2.Curve.generateKeyPair(),
        signedIdentityKey: identityKey,
        signedPreKey: crypto_2.signedKeyPair(identityKey, 1),
        registrationId: generics_1.generateRegistrationId(),
        advSecretKey: crypto_1.randomBytes(32).toString('base64'),
        nextPreKeyId: 1,
        firstUnuploadedPreKeyId: 1,
        serverHasPreKeys: false
    };
};
exports.initAuthCreds = initAuthCreds;
/** stores the full authentication state in a single JSON file */
const useSingleFileAuthState = (filename) => {
    // require fs here so that in case "fs" is not available -- the app does not crash
    const { readFileSync, writeFileSync, existsSync } = require('fs');
    let state = undefined;
    // save the authentication state to a file
    const saveState = () => {
        console.log('saving auth state');
        writeFileSync(filename, 
        // BufferJSON replacer utility saves buffers nicely
        JSON.stringify(state, generics_1.BufferJSON.replacer, 2));
    };
    if (existsSync(filename)) {
        const { creds, keys } = JSON.parse(readFileSync(filename, { encoding: 'utf-8' }), generics_1.BufferJSON.reviver);
        state = {
            creds: creds,
            // stores pre-keys, session & other keys in a JSON object
            // we deserialize it here
            keys: exports.initInMemoryKeyStore(keys, saveState)
        };
    }
    else {
        const creds = exports.initAuthCreds();
        const keys = exports.initInMemoryKeyStore({}, saveState);
        state = { creds: creds, keys: keys };
    }
    return { state, saveState };
};
exports.useSingleFileAuthState = useSingleFileAuthState;
